com.hazelcast.jet.core.TopologyChangeTest.when_coordinatorLeavesDuringExecution_then_jobCompletes

_master_ (commit https://github.com/hazelcast/hazelcast/commit/635391f382bd290648544b91b0b2ee106738d0e4)

Failed on Zulu JDK 16 - http://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-master-ZuluJDK16-nightly/6/testReport/com.hazelcast.jet.core/TopologyChangeTest/when_coordinatorLeavesDuringExecution_then_jobCompletes_liteMemberFlags_1__/

Stacktrace:
```
java.lang.AssertionError: expected:<2> but was:<0>
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotEquals(Assert.java:835)
	at org.junit.Assert.assertEquals(Assert.java:647)
	at org.junit.Assert.assertEquals(Assert.java:633)
	at com.hazelcast.jet.core.TopologyChangeTest.lambda$when_coordinatorLeavesDuringExecution_then_jobCompletes$4(TopologyChangeTest.java:299)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1260)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1362)
	at com.hazelcast.jet.core.TopologyChangeTest.when_coordinatorLeavesDuringExecution_then_jobCompletes(TopologyChangeTest.java:297)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:567)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:115)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:107)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.lang.Thread.run(Thread.java:831)```
```

Standard output:
```
Finished Running Test: when_nodeIsNotJobParticipant_then_initFails[liteMemberFlags(1)] in 1.037 seconds.
Started Running Test: when_coordinatorLeavesDuringExecution_then_jobCompletes[liteMemberFlags(1)]
02:31:14,656  INFO |when_coordinatorLeavesDuringExecution_then_jobCompletes[liteMemberFlags(1)]| - [MetricsConfigHelper] Time-limited test - [LOCAL] [dev] [5.0-SNAPSHOT] Overridden metrics configuration with system property 'hazelcast.metrics.collection.frequency'='1' -> 'MetricsConfig.collectionFrequencySeconds'='1'
02:31:14,657  INFO |when_coordinatorLeavesDuringExecution_then_jobCompletes[liteMemberFlags(1)]| - [logo] Time-limited test - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] 
	+       +  o    o     o     o---o o----o o      o---o     o     o----o o--o--o
	+ +   + +  |    |    / \       /  |      |     /         / \    |         |   
	+ + + + +  o----o   o   o     o   o----o |    o         o   o   o----o    |   
	+ +   + +  |    |  /     \   /    |      |     \       /     \       |    |   
	+       +  o    o o       o o---o o----o o----o o---o o       o o----o    o   
02:31:14,657  INFO |when_coordinatorLeavesDuringExecution_then_jobCompletes[liteMemberFlags(1)]| - [system] Time-limited test - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
02:31:14,658  INFO |when_coordinatorLeavesDuringExecution_then_jobCompletes[liteMemberFlags(1)]| - [system] Time-limited test - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Hazelcast Platform 5.0-SNAPSHOT (20210625 - 949deba) starting at [127.0.0.1]:5701
02:31:14,658  INFO |when_coordinatorLeavesDuringExecution_then_jobCompletes[liteMemberFlags(1)]| - [system] Time-limited test - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Cluster name: dev
02:31:14,660  INFO |when_coordinatorLeavesDuringExecution_then_jobCompletes[liteMemberFlags(1)]| - [MetricsConfigHelper] Time-limited test - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Collecting debug metrics and sending to diagnostics is enabled
02:31:14,663  WARN |when_coordinatorLeavesDuringExecution_then_jobCompletes[liteMemberFlags(1)]| - [CPSubsystem] Time-limited test - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] CP Subsystem is not enabled. CP data structures will operate in UNSAFE mode! Please note that UNSAFE mode will not provide strong consistency guarantees.
02:31:14,665  INFO |when_coordinatorLeavesDuringExecution_then_jobCompletes[liteMemberFlags(1)]| - [JetServiceBackend] Time-limited test - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Setting number of cooperative threads and default parallelism to 4
02:31:14,666  INFO |when_coordinatorLeavesDuringExecution_then_jobCompletes[liteMemberFlags(1)]| - [Diagnostics] Time-limited test - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Diagnostics disabled. To enable add -Dhazelcast.diagnostics.enabled=true to the JVM arguments.
02:31:14,666  INFO |when_coordinatorLeavesDuringExecution_then_jobCompletes[liteMemberFlags(1)]| - [LifecycleService] Time-limited test - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] [127.0.0.1]:5701 is STARTING
02:31:14,666  INFO |when_coordinatorLeavesDuringExecution_then_jobCompletes[liteMemberFlags(1)]| - [ClusterService] Time-limited test - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] 

Members {size:1, ver:1} [
	Member [127.0.0.1]:5701 - 5c1594a2-7911-4b39-9f48-3f9dfd08e435 this lite
]

02:31:14,666  INFO |when_coordinatorLeavesDuringExecution_then_jobCompletes[liteMemberFlags(1)]| - [JetExtension] Time-limited test - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Jet is enabled
02:31:14,666 DEBUG |when_coordinatorLeavesDuringExecution_then_jobCompletes[liteMemberFlags(1)]| - [JobCoordinationService] hz.fervent_euler.cached.thread-3 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
02:31:14,667  INFO |when_coordinatorLeavesDuringExecution_then_jobCompletes[liteMemberFlags(1)]| - [LifecycleService] Time-limited test - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] [127.0.0.1]:5701 is STARTED
02:31:14,667  INFO |when_coordinatorLeavesDuringExecution_then_jobCompletes[liteMemberFlags(1)]| - [MetricsConfigHelper] Time-limited test - [LOCAL] [dev] [5.0-SNAPSHOT] Overridden metrics configuration with system property 'hazelcast.metrics.collection.frequency'='1' -> 'MetricsConfig.collectionFrequencySeconds'='1'
02:31:14,667  INFO |when_coordinatorLeavesDuringExecution_then_jobCompletes[liteMemberFlags(1)]| - [logo] Time-limited test - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] 
	+       +  o    o     o     o---o o----o o      o---o     o     o----o o--o--o
	+ +   + +  |    |    / \       /  |      |     /         / \    |         |   
	+ + + + +  o----o   o   o     o   o----o |    o         o   o   o----o    |   
	+ +   + +  |    |  /     \   /    |      |     \       /     \       |    |   
	+       +  o    o o       o o---o o----o o----o o---o o       o o----o    o   
02:31:14,667  INFO |when_coordinatorLeavesDuringExecution_then_jobCompletes[liteMemberFlags(1)]| - [system] Time-limited test - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
02:31:14,667  INFO |when_coordinatorLeavesDuringExecution_then_jobCompletes[liteMemberFlags(1)]| - [system] Time-limited test - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] Hazelcast Platform 5.0-SNAPSHOT (20210625 - 949deba) starting at [127.0.0.1]:5702
02:31:14,667  INFO |when_coordinatorLeavesDuringExecution_then_jobCompletes[liteMemberFlags(1)]| - [system] Time-limited test - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] Cluster name: dev
02:31:14,668  INFO |when_coordinatorLeavesDuringExecution_then_jobCompletes[liteMemberFlags(1)]| - [HealthMonitor] hz.fervent_euler.HealthMonitor - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] processors=8, physical.memory.total=377.6G, physical.memory.free=185.5G, swap.space.total=4.0G, swap.space.free=4.0G, heap.memory.used=2.6G, heap.memory.free=1.4G, heap.memory.total=4.0G, heap.memory.max=4.0G, heap.memory.used/total=64.31%, heap.memory.used/max=64.31%, minor.gc.count=1326, minor.gc.time=13120ms, major.gc.count=0, major.gc.time=0ms, load.process=0.00%, load.system=100.00%, load.systemAverage=2.96, thread.count=65, thread.peakCount=2132, cluster.timeDiff=0, event.q.size=0, executor.q.async.size=0, executor.q.client.size=0, executor.q.client.query.size=0, executor.q.client.blocking.size=0, executor.q.query.size=0, executor.q.scheduled.size=0, executor.q.io.size=0, executor.q.system.size=0, executor.q.operations.size=0, executor.q.priorityOperation.size=0, operations.completed.count=0, executor.q.mapLoad.size=0, executor.q.mapLoadAllKeys.size=0, executor.q.cluster.size=0, executor.q.response.size=0, operations.running.count=0, operations.pending.invocations.percentage=0.00%, operations.pending.invocations.count=0, proxy.count=0, clientEndpoint.count=0, connection.active.count=0, client.connection.count=0, connection.count=0
02:31:14,668  INFO |when_coordinatorLeavesDuringExecution_then_jobCompletes[liteMemberFlags(1)]| - [MetricsConfigHelper] Time-limited test - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] Collecting debug metrics and sending to diagnostics is enabled
02:31:14,670  WARN |when_coordinatorLeavesDuringExecution_then_jobCompletes[liteMemberFlags(1)]| - [CPSubsystem] Time-limited test - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] CP Subsystem is not enabled. CP data structures will operate in UNSAFE mode! Please note that UNSAFE mode will not provide strong consistency guarantees.
02:31:14,672  INFO |when_coordinatorLeavesDuringExecution_then_jobCompletes[liteMemberFlags(1)]| - [JetServiceBackend] Time-limited test - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] Setting number of cooperative threads and default parallelism to 4
02:31:14,673  INFO |when_coordinatorLeavesDuringExecution_then_jobCompletes[liteMemberFlags(1)]| - [Diagnostics] Time-limited test - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] Diagnostics disabled. To enable add -Dhazelcast.diagnostics.enabled=true to the JVM arguments.
02:31:14,673  INFO |when_coordinatorLeavesDuringExecution_then_jobCompletes[liteMemberFlags(1)]| - [LifecycleService] Time-limited test - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] [127.0.0.1]:5702 is STARTING
02:31:14,674  INFO |when_coordinatorLeavesDuringExecution_then_jobCompletes[liteMemberFlags(1)]| - [MockServer] Time-limited test - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] Created connection to endpoint: [127.0.0.1]:5701, connection: MockConnection{localEndpoint=[127.0.0.1]:5702, remoteEndpoint=[127.0.0.1]:5701, alive=true}
02:31:14,674  INFO |when_coordinatorLeavesDuringExecution_then_jobCompletes[liteMemberFlags(1)]| - [MockServer] hz.fervent_euler.priority-generic-operation.thread-0 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Created connection to endpoint: [127.0.0.1]:5702, connection: MockConnection{localEndpoint=[127.0.0.1]:5701, remoteEndpoint=[127.0.0.1]:5702, alive=true}
02:31:14,674  INFO |when_coordinatorLeavesDuringExecution_then_jobCompletes[liteMemberFlags(1)]| - [ClusterService] hz.fervent_euler.priority-generic-operation.thread-0 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] 

Members {size:2, ver:2} [
	Member [127.0.0.1]:5701 - 5c1594a2-7911-4b39-9f48-3f9dfd08e435 this lite
	Member [127.0.0.1]:5702 - 11b28e93-2a10-4eb2-992f-86d370186ffd
]

02:31:14,767 DEBUG |when_coordinatorLeavesDuringExecution_then_jobCompletes[liteMemberFlags(1)]| - [JobCoordinationService] hz.fervent_euler.cached.thread-3 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
02:31:14,774  INFO |when_coordinatorLeavesDuringExecution_then_jobCompletes[liteMemberFlags(1)]| - [ClusterService] hz.peaceful_euler.priority-generic-operation.thread-0 - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] 

Members {size:2, ver:2} [
	Member [127.0.0.1]:5701 - 5c1594a2-7911-4b39-9f48-3f9dfd08e435 lite
	Member [127.0.0.1]:5702 - 11b28e93-2a10-4eb2-992f-86d370186ffd this
]

02:31:14,867 DEBUG |when_coordinatorLeavesDuringExecution_then_jobCompletes[liteMemberFlags(1)]| - [JobCoordinationService] hz.fervent_euler.cached.thread-3 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
02:31:14,967 DEBUG |when_coordinatorLeavesDuringExecution_then_jobCompletes[liteMemberFlags(1)]| - [JobCoordinationService] hz.fervent_euler.cached.thread-3 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
02:31:15,067 DEBUG |when_coordinatorLeavesDuringExecution_then_jobCompletes[liteMemberFlags(1)]| - [JobCoordinationService] hz.fervent_euler.cached.thread-3 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
02:31:15,167 DEBUG |when_coordinatorLeavesDuringExecution_then_jobCompletes[liteMemberFlags(1)]| - [JobCoordinationService] hz.fervent_euler.cached.thread-4 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
02:31:15,174  INFO |when_coordinatorLeavesDuringExecution_then_jobCompletes[liteMemberFlags(1)]| - [JetExtension] Time-limited test - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] Jet is enabled
02:31:15,174  INFO |when_coordinatorLeavesDuringExecution_then_jobCompletes[liteMemberFlags(1)]| - [LifecycleService] Time-limited test - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] [127.0.0.1]:5702 is STARTED```
```

--- a/hazelcast/src/test/java/com/hazelcast/jet/core/TopologyChangeTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/jet/core/TopologyChangeTest.java
@@ -272,8 +272,15 @@ public class TopologyChangeTest extends JetTestSupport {
             NoOutputSourceP.executionStarted.await();
 
             instances[0].getLifecycleService().terminate();
+            // Wait for job executions terminated in non-terminated instances
+            // before proceeding. Otherwise, job executions on these members
+            // may complete successfully without exception and without calling
+            // Processor#close.
+            for (int i = 1; i < 3; i++) {
+                JetServiceBackend jetServiceBackend = getJetServiceBackend(instances[i]);
+                jetServiceBackend.getJobExecutionService().waitAllExecutionsTerminated();
+            }
             NoOutputSourceP.proceedLatch.countDown();
-
             future.get();
             fail();
         } catch (ExecutionException expected) {

--- a/hazelcast/src/test/java/com/hazelcast/jet/core/TopologyChangeTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/jet/core/TopologyChangeTest.java
@@ -276,7 +276,7 @@ public class TopologyChangeTest extends JetTestSupport {
             // before proceeding. Otherwise, job executions on these members
             // may complete successfully without exception and without calling
             // Processor#close.
-            for (int i = 1; i < 3; i++) {
+            for (int i = 1; i < instances.length; i++) {
                 JetServiceBackend jetServiceBackend = getJetServiceBackend(instances[i]);
                 jetServiceBackend.getJobExecutionService().waitAllExecutionsTerminated();
             }

